### PR TITLE
treewide: completely remove types.loaOf

### PR DESCRIPTION
**Pull request by rnhmjoj**
###### Motivation for this change
Finalise the removal of loaOf, started with #63103.
See also https://github.com/NixOS/nixpkgs/issues/77189#issuecomment-583927360.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Review by worldofpeace — CHANGES_REQUESTED** [review]
Why don't we turn the warning in the type into something fatal to help people out? 
Also mention in the release notes that it is now removed entirely?

**Comment by rnhmjoj** [comment]
Thank you, done. I don't know how the Nixos options integration in docbook happen, so is there a file that needs to be updated for the removal? I couldn't find any more references to `loaOf`.

**Comment by Infinisil** [comment]
No idea either, haven't looked into it

**Comment by jtojnar** [comment]
We replace the `?` in `id`s for `_` here:

https://github.com/NixOS/nixpkgs/blob/master/nixos/lib/make-options-doc/options-to-docbook.xsl#L23

The id just comes from the option name itself IIRC.

**Comment by rnhmjoj** [comment]
@jtojnar, sorry but I understand very little of what's going on in that file. I still don't see anything related to loaOf in it, what should be changed?

**Comment by jtojnar** [comment]
We can drop the outermost `replace` if no other type adds `?` to the generated option name lika `loaOf` did in `getSubOptions`.

**Comment by rnhmjoj** [comment]
Ok, I think I get it. It's because `loaOf` uses `<name?>` as a placeholder.
Yes, it seems no other type does this and the manual builds, so it should be safe to remove.

**Review by infinisil — APPROVED** [review]
LGTM

File: lib/types.nix [hunk 1]
```diff
@@ -326,110 +326,15 @@ rec {
...
-        substSubModules = m: loaOf (elemType.substSubModules m);
-        functor = (defaultFunctor name) // { wrapped = elemType; };
-      };
+    # drop this in the future:
+    loaOf =
+      let msg =
+        ''
+          `types.loaOf` is deprecated and mixing list with attribute values
+          is no longer possible; use only `types.attrsOf` instead.
+          See https://git.io/JUvd1 for the motivation.
```
Comment (cole-h) [inline lib/types.nix:335]:
```suggestion
          `types.loaOf` has been removed and mixing lists with attribute values
          is no longer possible; please use `types.attrsOf` instead.
          See https://git.io/JUvd1 for the motivation.
```

Even though `types.loaOf` is aliased to `types.attrsOf` now, I still think this counts as being removed. However, if this wording isn't desired, I think at least the following minor changes should be included:

```suggestion
          `types.loaOf` is deprecated and mixing lists with attribute values
          is no longer possible; please use `types.attrsOf` instead.
          See https://git.io/JUvd1 for the motivation.
```

Comment (rnhmjoj) [reply lib/types.nix]: Thanks for the suggestions. The rewording looks good to me: I've implemented it.

Comment (jtojnar) [reply lib/types.nix]: Could we not use URL shorteners? They obscure the target location and in this case the URL is not even that long.

File: lib/types.nix [hunk 2]
```diff
@@ -326,110 +326,15 @@ rec {
...
-        description = "list or attribute set of ${elemType.description}s";
-        check = x: isList x || isAttrs x;
-        merge = loc: defs: attrOnly.merge loc (convertAllLists loc defs);
-        emptyValue = { value = {}; };
-        getSubOptions = prefix: elemType.getSubOptions (prefix ++ ["<name?>"]);
-        getSubModules = elemType.getSubModules;
-        substSubModules = m: loaOf (elemType.substSubModules m);
-        functor = (defaultFunctor name) // { wrapped = elemType; };
-      };
+    # drop this in the future:
```
Comment (cole-h) [inline lib/types.nix:329]: I think this line should stand out more (sorry for another suggestion). Maybe just add a `FIXME:` or `TODO:` prefix, so it's easier to notice that something still needs doing here?

File: lib/types.nix (hunk 1 above)
Comment (rnhmjoj) [reply lib/types.nix]: Switched to original link.

File: lib/types.nix (hunk 2 above)
Comment (rnhmjoj) [reply lib/types.nix]: Done, I also applied the changes to `list`, which shared a similar fate.

**Review by cole-h — APPROVED** [review]
Diff LGTM, thanks!

**Review by worldofpeace — APPROVED** [review]

**Comment by rnhmjoj** [comment]
Thank you all!

**Comment by worldofpeace** [comment]
Absolutely, and I return the thanks ten fold :sparkles:

**Pull request merged**